fdb-record-layer-lucene/src/main/java/com/apple/foundationdb/record/lucene/codec/LuceneOptimizedCodec.java

```diff
@@ -0,0 +1,126 @@
+/*
+ * LuceneOptimizedCodec.java
+ *
+ * This source file is part of the FoundationDB open source project
+ *
+ * Copyright 2015-2021 Apple Inc. and the FoundationDB project authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.apple.foundationdb.record.lucene.codec;
+
+import org.apache.lucene.codecs.Codec;
+import org.apache.lucene.codecs.CompoundFormat;
+import org.apache.lucene.codecs.DocValuesFormat;
+import org.apache.lucene.codecs.FieldInfosFormat;
+import org.apache.lucene.codecs.LiveDocsFormat;
+import org.apache.lucene.codecs.NormsFormat;
+import org.apache.lucene.codecs.PointsFormat;
+import org.apache.lucene.codecs.PostingsFormat;
+import org.apache.lucene.codecs.SegmentInfoFormat;
+import org.apache.lucene.codecs.StoredFieldsFormat;
+import org.apache.lucene.codecs.TermVectorsFormat;
+import org.apache.lucene.codecs.lucene50.Lucene50StoredFieldsFormat;
+import org.apache.lucene.codecs.lucene70.Lucene70Codec;
+
+/**
+ *
+ * Codec with a few optimizations for speeding up compound files
+ * sitting on FoundationDB.
```

Comment:
I'm not 100% sure what the right place to document this would be, but if there are some kind of special backwards- or forwards-compatibility concerns with this codec, it might be good to note them. For example, can data written with this codec be read by un-upgraded code using an older codec (and vice versa)?

Reply:
If I understand correctly, backwards or forwards compatibility is not supported. We may have to manually rebuild the indexes for the existing data before handling search requests. Please correct me if I am wrong. @jleach4

Reply:
I see. If that's the case, we should probably document that somewhere, if only in a `breaking change` in the release note so that people know they need to rebuild any Lucene indexes they have. Normally, I'd argue for some kind of transition scheme without dataloss, but that's probably acceptable for an experimental index like this (as long as we communicate what we're doing).

Reply:
Added 2 lines in the javadoc: ```Forwards/backwards compatibility is not supported during transition from the {@link Lucene70Codec} to this implementation unfortunately. Indexes need to rebuilt before handling search requests.```

Reply:
Okay, that's probably good, though I still think we should include it in the release notes (as people upgrading are theoretically more likely to check there than the javadoc comments).

Reply:
@alecgrieser I agree their needs to be a compatibility story.  I also agree with the release notes piece as well.